FaceDetectionServer/client/multi_client.py

```diff
@@ -0,0 +1,437 @@
+
+class Client:
+    """ Base Client class """
+
+    # Initialize "Grand total" class variables.
+    total_latency_full_process = 0
+    count_latency_full_process = 0
+    total_latency_network_only = 0
+    count_latency_network_only = 0
+    total_server_processing_time = 0
+    count_server_processing_time = 0
+
+    def __init__(self, host, port):
+        # Initialize instance variables.
+        self.host = host
+        self.port = port
+        self.do_server_stats = False
+        self.show_responses = False
+        self.total_latency_full_process = 0
+        self.count_latency_full_process = 0
+        self.total_latency_network_only = 0
+        self.count_latency_network_only = 0
+        self.total_server_processing_time = 0
+        self.count_server_processing_time = 0
+        self.image_file_name = None
+        self.latency_start_time = 0
+        self.loop_count = 0
+        self.num_repeat = 0
+        self.filename_list = []
+        self.filename_list_index = 0
+        self.json_params = None
+        self.base64 = False
+        logger.debug("host:port = %s:%d" %(self.host, self.port))
+
+    def start(self):
+        logger.debug("image file(s) %s %s" %(self.image_file_name, self.filename_list))
+
+    def next_file_name(self):
+        """ If the filename_list array is has more than 1, get the next value. """
+        if len(self.filename_list) > 1:
+            self.filename_list_index += 1
+            if self.filename_list_index >= len(self.filename_list):
+                self.filename_list_index = 0
+            self.image_file_name = self.filename_list[self.filename_list_index]
+
+    def time_open_socket(self):
+        now = time.time()
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM);
+        sock.settimeout(2)
+        result = sock.connect_ex((self.host, self.port))
+        if result != 0:
+            logger.error("Could not connect to %s on port %d" %(self.host, self.port))
+            return
+        millis = (time.time() - now)*1000
+        elapsed = "%.3f" %millis
+        logger.info("%s ms to open socket" %(elapsed))
+        self.total_latency_network_only = self.total_latency_network_only + millis
+        self.count_latency_network_only += 1
+
+    def icmp_ping(self):
+        args=[PING, '-c', '1', '-W', '1', self.host]
+        p_ping = subprocess.Popen(args,
+                                  shell=False,
+                                  stdout=subprocess.PIPE)
+        # save ping stdout
+        p_ping_out = str(p_ping.communicate()[0])
+
+        if (p_ping.wait() == 0):
+            logger.info(p_ping_out)
+            # rtt min/avg/max/mdev = 61.994/61.994/61.994/0.000 ms
+            search = re.search(PING_REGEX, p_ping_out, re.M|re.I)
+            ping_rtt = search.group(2)
+            logger.info("%s ms ICMP ping" %(ping_rtt))
+            self.total_latency_network_only = self.total_latency_network_only + float(ping_rtt)
```

Comment:
Do we ever add Ping results with Opening and closing socket results (they both modify total_latency_network_only)? Or are they separate cases?

Reply:
It's either/or. Which is used is determined by the "-n", "--network-latency" parameter, with SOCKET being the default. Basically:
```
    if self.net_latency_method == "SOCKET":
        self.time_open_socket()
    else:
        self.icmp_ping()
```